src/hypothesis/strategies.py

```diff
@@ -836,31 +836,80 @@ def dm_func(denom):
 
 @cacheable
 @defines_strategy
-def decimals(
-    min_value=None, max_value=None, allow_nan=None, allow_infinity=None
-):
-    """Generates instances of decimals.Decimal.
+def decimals(min_value=None, max_value=None,
+             allow_nan=True, allow_infinity=None, places=None):
+    """Generates instances of decimals.Decimal, which may be:
 
-    - If min_value is not None, all values will be >= min_value.
-    - If max_value is not None, all values will be <= max_value.
-    - If min_value or max_value is not None, it is an error to enable
-      allow_nan.
-    - If both min_value and max_value are not None, it is an error to enable
-      allow_infinity.
+    - A finite rational number, between ``min_value`` and ``max_value``.
+    - Not a Number, if ``allow_nan`` is truthy.
+    - Positive or negative infinity, if ``max_value`` and ``min_value``
+      respectively are None, and ``allow_infinity`` is not False.  None means
+      "allow infinity, unless excluded by the min and max values".
 
-    Where not explicitly ruled out by the bounds, all of infinity, -infinity
-    and NaN are possible values generated by this strategy.
+    Note that where floats have one `NaN` value, Decimals have four: signed,
+    and either *quiet* or *signalling*.  See `the decimal module docs
+    <https://docs.python.org/3/library/decimal.html#special-values>`_ for
+    more information on special values.
+
+    If ``places`` is not None, all finite values drawn from the strategy will
+    have that number of digits after the decimal place.
 
     """
-    floats_strategy = floats(
-        min_value=min_value, max_value=max_value,
-        allow_nan=allow_nan, allow_infinity=allow_infinity)
-    fractions_strategy = fractions(min_value=min_value, max_value=max_value)
-
-    return (
-        floats_strategy.map(float_to_decimal) |
-        fractions_strategy.map(lambda f: Decimal(f.numerator) / f.denominator)
-    )
+    # Convert min_value and max_value to Decimal values, and validate args
+    check_valid_integer(places)
+    if places is not None and places < 0:
+        raise InvalidArgument('places=%r may not be negative' % places)
+
+    if min_value is not None:
+        min_value = Decimal(min_value)
+        if min_value.is_infinite() and min_value < 0:
+            min_value = None
+        elif not min_value.is_finite():
+            raise InvalidArgument(u'Invalid min_value=%r' % min_value)
+    if max_value is not None:
+        max_value = Decimal(max_value)
+        if max_value.is_infinite() and max_value > 0:
+            max_value = None
+        elif not max_value.is_finite():
+            raise InvalidArgument(u'Invalid max_value=%r' % max_value)
+    check_valid_bound(min_value, 'min_value')
+    check_valid_bound(max_value, 'max_value')
+    check_valid_interval(min_value, max_value, 'min_value', 'max_value')
+
+    if allow_infinity is None:
+        allow_infinity = (min_value is None) or (max_value is None)
```

Comment:
I'm still puzzled why you summarize here, only to deconstruct further on.

New suggestion, have just:

```
if allow_infinity and (None not in (min_value, max_value)):
   raise ...
```

here, and at line 907:

```
if (allow_infinity is None and min_value is None) or allow_infinity:
    special.append(Decimal(-Infinity))
if (allow_infinity is None and max_value is None) or allow_infinity:
    special.append(Decimal(Infinity))
```

Also, there might be an error condition for ```allow_infinity``` = ```False``` which is missing, if, e.g., the ```max_value``` or ```min_value``` are specified as infinity or -infinity.